### PR TITLE
Add go and java pipeline for +Add flows

### PR DESCRIPTION
**Pull request by pradeepitm12**
This adds pipelines for go and java based source.
Pipeline consist of build and deploy tasks
required for +Add flows in console

Signed-off-by: Pradeep Kumar <pradkuma@redhat.com>

### CODE REVIEW

**Comment by pradeepitm12** [comment]
@nikhil-thomas @sthaha 
This is an initial PR from a series to add pipeline template for most used languages.
Please let me know if we the directory structures and the version is handled properly.

**Comment by pradeepitm12** [comment]
cc @hrishin

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
+---
```
Comment (sthaha) [inline deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: @pradeepitm12 why is this added to `0.8.0` ?

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+          - name: source
+            resource: app-source
+        outputs:
+          - name: image
+            resource: app-image
+      params:
+        - name: TLSVERIFY
+          value: "false"
+
+    - name: deploy
```
Comment (siamaksade) [inline deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: `runAfter` is missing in the pipelines in this PR so that deploy runs after build

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml (hunk 1 above)
Comment (nikhil-thomas) [reply deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: +1
@pradeepitm12 could you move this to `deploy/resources/v0.9.2/addons`

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+---
+apiVersion: tekton.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: s2i-go-build-and-deploy-pipeline
+  namespace: openshift
```
Comment (nikhil-thomas) [inline deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: @pradeepitm12 
is the `namespace: openshift` a hard constraint for this feature to work in ui.

if no, could you remove metadata.name from the pipeline samples.

if yes, this pr will need to handle that in the controller logic.

**Comment by nikhil-thomas** [comment]
@pradeepitm12 
nit: could you add a `-` in versioned names of pipeline templates

eg: change 
`s2i-gov0-8-0-build-deploy.yaml`
to
`s2i-go-v0-8-0-build-deploy.yaml`

like the clustertask versioned names: https://github.com/openshift/tektoncd-pipeline-operator/tree/master/deploy/resources/v0.9.2/addons/clustertasks/s2i-go

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml (hunk 3 above)
Comment (pradeepitm12) [reply deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: Yes, for the time being, `namespace: openshift` is a hard constraint.

**Comment by nikhil-thomas** [comment]
@pradeepitm12 @vdemeester 
as per our discussion, lets add an annotation to the pipeline samples to prevent the operator from replacing `namespace: openshift` with `namespace: openshift-pipelines`

something like: `operator.tekton.dev/preserve-namespace: true`

the controller logic has to be changed to respect the annotation.

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml (hunk 2 above)
Comment (sthaha) [reply deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: 👏good catch!

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-gov0-8-0-build-deploy.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
+---
+apiVersion: tekton.dev/v1alpha1
```
Comment (sthaha) [inline deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-gov0-8-0-build-deploy.yaml]: * could you please name the dir `pipelines` 
* and follow the same convention used in clustertasks ?
* why do we need a versioned pipeline?

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+---
+apiVersion: tekton.dev/v1alpha1
+kind: Pipeline
+metadata:
+  name: s2i-go-build-and-deploy-pipeline
```
Comment (sthaha) [inline deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: should it end with `-pipeline`? isn't the `kind` already `Pipeline`?

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-java-11/s2i-java-11-build-deploy.yaml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        - name: TLSVERIFY
+          value: "false"
+
+    - name: deploy
+      taskRef:
+        name: openshift-client
+        kind: ClusterTask
+      params:
+        - name: ARGS
+          value:
```
Comment (sthaha) [inline deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-java-11/s2i-java-11-build-deploy.yaml]: nit: `value: [ new-app, --docker-image, $(params.IMAGE_NAME) ] ` may be easier to read

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-gov0-8-0-build-deploy.yaml (hunk 4 above)
Comment (pradeepitm12) [reply deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-gov0-8-0-build-deploy.yaml]: Changed to pipelines,
After testing with the console, it makes sense to have non-versioned pipeline only.

File: deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml (hunk 2 above)
Comment (pradeepitm12) [reply deploy/resources/v0.8.0/addons/pipelinetemplates/s2i-go/s2i-go-build-deploy.yaml]: Good catch, updated.

File: pkg/flag/flag.go [hunk 7]
```diff
@@ -28,6 +28,8 @@ const (
 	TriggerControllerName = "tekton-triggers-controller"
 	TriggerWebhookName    = "tekton-triggers-webhook"
 
+	PreserveNamespace = "operator.tekton.dev/preserve-namespace"
```
Comment (nikhil-thomas) [inline pkg/flag/flag.go]: nit: How about renaming the const to `PreserceNSAnnotation"

File: pkg/utils/transform/transform.go [hunk 8]
```diff
@@ -35,3 +35,15 @@ func InjectDefaultSA(defaultSA string) mf.Transformer {
 		return nil
 	}
 }
+
+func InjectNamespaceCustom(preserveNamespace string, tf mf.Transformer) mf.Transformer {
+	return func(u *unstructured.Unstructured) error {
```
Comment (nikhil-thomas) [inline pkg/utils/transform/transform.go]: as we know the `transformer` passed is always `manifetival.InjectNamespace`, we can eliminate the second argument.

before the return statement, let's add 
```
func InjectNamespaceCustom(preserveNamespace string) mf.Transformer {
    tf := mf.InjectNamespace(preserveNamespace)
    return func(u *unstructured.Unstructured) error {
    ...
    ...
```
then use tf in the returned function as it is.

Comment (pradeepitm12) [reply pkg/utils/transform/transform.go]: Here TargetNamespace is required which comes from `cfg.Spec.TargetNamespace`

Comment (nikhil-thomas) [reply pkg/utils/transform/transform.go]: i see, then can we pass the `targetNamespace` as the second argument to `InjectNamespaceCustom`.

and construct the namespace transformer inside InjectNamespaceCustom.
something like
```
func InjectNamespaceCustom(preserveNamespace, targetNamespace string) mf.Transformer {
    tf := mf.InjectNamespace(targetNamespace)
    return func(u *unstructured.Unstructured) error {
    ...
    ...
```
